feat(turbo): task access trace & cached configs

tknickman opened the pull request:
### Description

v0 of work required for zero config (experimental):
1. Support config caching
2. Support task tracing

Requires: https://github.com/vercel/next.js/pull/59553

#### TODO
- [ ] Testing
- [ ] Support caching outputs on first run instead of second (turbo.json uncoupling)

chris-olszewski commented on crates/turborepo-lib/src/run/cache.rs:
```diff
@@ -374,3 +382,88 @@ impl TaskCache {
+    }
+
+    pub async fn save(&self) -> Result<(), CacheError> {
+        match self.exists() {
```
There's a TOCTOU race condition where the file gets cleaned up between us checking for it's existence and it actually getting read into memory.

chris-olszewski replied:
Tracing through this, I don't think there's an easy way around it unless we rework the cache abstraction a fair bit. I think it's fine to leave as is 🤷

chris-olszewski commented on crates/turborepo-lib/src/run/task_access.rs:
```diff
@@ -0,0 +1,244 @@
+    }
+
+    // read package.json at root
+    let package_json_path = repo_root.join_components(&["package.json"]);
```
Is this feature only intended to be used in single package workspaces?

tknickman replied:
For now, yes (but we'll roll this out more broadly soon)

chris-olszewski commented on crates/turborepo-lib/src/task_graph/visitor.rs:
```diff
@@ -838,21 +860,49 @@ impl ExecContext {
-                        self.task_id.clone(),
-                        self.task_cache.expanded_outputs().to_vec(),
-                    );
+                    let mut cacheable = true;
+                    // If stdout_writer flush is successful, proceed with task processing
+                    if self.task_access.is_enabled() {
+                        // Task access enabled: Handle trace file processing
+                        if let Some(trace) =
+                            TaskAccessTraceFile::read(&self.task_access.repo_root, &self.task_hash)
+                        {
+                            // Trace file found
+                            if trace.can_cache(&self.task_access.repo_root) {
+                                // Cache trace and save task outputs if trace is cacheable
+                                self.task_access
+                                    .save_trace(trace, self.task_id_for_display.to_string());
+                            } else {
+                                cacheable = false;
+                            }
+                        }
+                    } else {
+                        // Task access not enabled: Directly save task outputs
+                    }
+
+                    if cacheable {
+                        // Attempt to save task outputs and log any errors encountered
+                        if let Err(e) = self
+                            .task_cache
+                            .save_outputs(&mut prefixed_ui, task_duration, telemetry)
+                            .await
+                        {
+                            error!("error caching output: {e}");
+                        } else {
+                            // If no errors, update hash tracker with expanded outputs
+                            self.hash_tracker.insert_expanded_outputs(
+                                self.task_id.clone(),
+                                self.task_cache.expanded_outputs().to_vec(),
+                            );
+                        }
+                    }
```
This "can we cache" logic feels like it should belong totally in the task access module:
```
    // Whether we can cache the given task, returning None if task access isn't
    // enabled or the trace can't be found
    pub fn can_cache(&self, task_hash: &str, task_id: &str) -> Option<bool> {
        if !self.is_enabled() {
            return None;
        }
        let trace = TaskAccessTraceFile::read(&self.repo_root, task_hash)?;
        if trace.can_cache(&self.repo_root) {
            self.save_trace(trace, task_id.to_string());
            Some(true)
        } else {
            Some(false)
        }
    }
```
Which simplifies the changes in Visitor to
```suggestion
                   if self
                        .task_access
                        .can_cache(&self.task_hash, &self.task_id_for_display)
                        .unwrap_or(true)
                    {
                        if let Err(e) = self
                            .task_cache
                            .save_outputs(&mut prefixed_ui, task_duration, telemetry)
                            .await
                        {
                            error!("error caching output: {e}");
                        } else {
                            // If no errors, update hash tracker with expanded outputs
                            self.hash_tracker.insert_expanded_outputs(
                                self.task_id.clone(),
                                self.task_cache.expanded_outputs().to_vec(),
                            );
                        }
                    }
```

chris-olszewski commented on crates/turborepo-lib/src/task_graph/visitor.rs:
```diff
@@ -838,21 +860,49 @@ impl ExecContext {
+                            if trace.can_cache(&self.task_access.repo_root) {
+                                // Cache trace and save task outputs if trace is cacheable
+                                self.task_access
+                                    .save_trace(trace, self.task_id_for_display.to_string());
```
Just making sure my understanding is correct: we want this to ensure we're getting a valid `TaskName` that can appear in a pipeline definition so we're using `task_id_for_display` as it'll end up being a valid pipeline task name?

tknickman replied:
Yep that's correct, a little weird in this case, but we need that to build the pipeline later on.

chris-olszewski reviewed (changes requested):
Mostly concerned about how this works with multi-package workspaces.

tknickman commented on crates/turborepo-lib/src/run/cache.rs:
```diff
@@ -32,11 +36,15 @@ pub enum Error {
 pub struct RunCache {
     task_output_mode: Option<OutputLogsMode>,
-    cache: AsyncCache,
+    cache: Arc<AsyncCache>,
```
Good call, broke out that PR into it's own: https://github.com/vercel/turbo/pull/7293

tknickman commented:
@chris-olszewski PR updated.

There were a lot of conflicts here, so split the commits between resolving conflicts, and addressing review. But let me know if you have any questions!

NOTE: This is stacked on top of https://github.com/vercel/turbo/pull/7293

chris-olszewski reviewed (approved):
LFG 🚀
